File: core-render-ohos/src/main/cpp/libohos_render/expand/modules/sftp/KRSftpConnectionModule.cpp
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "KRSftpConnectionModule.h"

#include <algorithm>
#include <thread>
#include <vector>

#include "SftpErrorFormatter.h"
#include "SftpJsonStore.h"
#include "libohos_render/export/IKRRenderViewExport.h"
#include "libohos_render/foundation/type/KRRenderValue.h"

namespace kuikly {
namespace module {

namespace {

constexpr const char *kConnectionsFile = "sftp_connections.json";
constexpr const char *kFavoritesFile = "sftp_favorites.json";
constexpr const char *kHistoryFile = "sftp_playback_history.json";

std::string MakeConnectionId(const cJSON *item) {
    std::string host = sftp_store::CJsonString(item, "host");
    long long port = sftp_store::CJsonNumber(item, "port", 22);
    std::string user = sftp_store::CJsonString(item, "user");
    return host + ":" + std::to_string(port) + "/" + user + "#" +
           std::to_string(sftp_store::NowMs());
}

/** label 重复时追加 (2)/(3)…，与各端行为一致。 */
std::string UniqueLabel(const cJSON *array, const std::string &label) {
    bool dup = true;
    std::string candidate = label;
    int suffix = 2;
    while (dup) {
        dup = false;
        for (const cJSON *e = array ? array->child : nullptr; e != nullptr; e = e->next) {
            if (sftp_store::CJsonString(e, "label") == candidate) {
                dup = true;
                break;
            }
        }
        if (dup) {
            candidate = label + "(" + std::to_string(suffix++) + ")";
        }
    }
    return candidate;
}

}  // namespace

// MODULE_NAME 在头文件里已是 `static constexpr const char *`（C++17 隐式 inline），无需在此定义。
const char KRSftpConnectionModule::METHOD_ADD[]             = "add";
const char KRSftpConnectionModule::METHOD_UPDATE[]         = "update";
const char KRSftpConnectionModule::METHOD_REMOVE[]          = "remove";
const char KRSftpConnectionModule::METHOD_LIST[]            = "list";
const char KRSftpConnectionModule::METHOD_GET[]             = "get";
const char KRSftpConnectionModule::METHOD_TOUCH_LAST_USED[] = "touchLastUsed";

KRAnyValue KRSftpConnectionModule::CallMethod(bool sync, const std::string &method, KRAnyValue params,
                                                const KRRenderCallback &callback) {
    std::string filesDir;
    if (auto root = GetRootView().lock()) {
        filesDir = root->GetContext()->Config()->GetFilesDir();
    }
    std::thread([method, params, callback, filesDir]() {
        try {
            if (filesDir.empty()) {
                throw std::runtime_error("connection store: files dir unavailable");
            }
            sftp_store::JsonArrayFile store(
                sftp_store::PathIn(filesDir, kConnectionsFile), "id");

            if (method == METHOD_ADD) {
                std::string json = params ? params->toString() : "";
                cJSON *item = cJSON_Parse(json.c_str());
                if (!item) {
                    throw std::runtime_error("connection add: invalid json");
                }
                cJSON *array = store.Load();
                std::string id = sftp_store::CJsonString(item, "id");
                if (id.empty()) {
                    id = MakeConnectionId(item);
                    sftp_store::CJsonSetString(item, "id", id);
                }
                long long now = sftp_store::NowMs();
                if (sftp_store::CJsonNumber(item, "createdAt", 0) == 0) {
                    sftp_store::CJsonSetNumber(item, "createdAt", now);
                }
                sftp_store::CJsonSetString(
                    item, "label", UniqueLabel(array, sftp_store::CJsonString(item, "label")));
                store.Upsert(array, item);
                bool ok = store.SaveAndFree(array);
                cJSON_Delete(item);
                if (!ok) {
                    throw std::runtime_error("connection add: save failed");
                }
                KRRenderValueMap result;
                result["id"] = KRRenderValue::Make(id);
                callback(KRRenderValue::Make(result));
                return;
            }

            if (method == METHOD_UPDATE) {
                std::string json = params ? params->toString() : "";
                cJSON *item = cJSON_Parse(json.c_str());
                if (!item) {
                    throw std::runtime_error("connection update: invalid json");
                }
                cJSON *array = store.Load();
                std::string id = sftp_store::CJsonString(item, "id");
                cJSON *existing = store.FindById(array, id);
                if (existing) {
                    // 保留 createdAt，其余字段覆盖。
                    sftp_store::CJsonSetNumber(
                        item, "createdAt", sftp_store::CJsonNumber(existing, "createdAt", 0));
                    store.Upsert(array, item);
                } else {
                    store.Upsert(array, item);
                }
                bool ok = store.SaveAndFree(array);
                cJSON_Delete(item);
                if (!ok) {
                    throw std::runtime_error("connection update: save failed");
                }
                callback(KRRenderValue::Make(KRRenderValueMap{}));
                return;
            }

            if (method == METHOD_REMOVE) {
                auto map = params ? params->toMap() : KRRenderValue::Map{};
                auto it = map.find("id");
                std::string id = (it != map.end() && it->second) ? it->second->toString() : "";
                cJSON *array = store.Load();
                store.RemoveById(array, id);
                store.SaveAndFree(array);
                // 级联清理收藏与播放历史（§21.4.5）。
                if (!id.empty()) {
                    sftp_store::JsonArrayFile favorites(
                        sftp_store::PathIn(filesDir, kFavoritesFile), "id");
                    cJSON *favArray = favorites.Load();
                    if (favorites.RemoveByField(favArray, "connectionId", id) > 0) {
                        favorites.SaveAndFree(favArray);
                    } else {
                        cJSON_Delete(favArray);
                    }
                    sftp_store::JsonArrayFile history(
                        sftp_store::PathIn(filesDir, kHistoryFile), "id");
                    cJSON *hisArray = history.Load();
                    if (history.RemoveByField(hisArray, "connectionId", id) > 0) {
                        history.SaveAndFree(hisArray);
                    } else {
                        cJSON_Delete(hisArray);
                    }
                }
                callback(KRRenderValue::Make(KRRenderValueMap{}));
                return;
            }

            if (method == METHOD_LIST) {
                cJSON *array = store.Load();
                std::vector<cJSON *> items;
                for (cJSON *e = array->child; e != nullptr; e = e->next) {
                    items.push_back(e);
                }
                std::stable_sort(items.begin(), items.end(), [](const cJSON *a, const cJSON *b) {
                    return sftp_store::CJsonNumber(a, "lastUsedAt", 0) >
                           sftp_store::CJsonNumber(b, "lastUsedAt", 0);
                });
                KRRenderValue::Array out;
                for (const cJSON *e : items) {
                    out.push_back(sftp_store::FromCJson(e));
                }
                cJSON_Delete(array);
                KRRenderValueMap result;
                result["items"] = KRRenderValue::Make(out);
                callback(KRRenderValue::Make(result));
                return;
            }

            if (method == METHOD_GET) {
                auto map = params ? params->toMap() : KRRenderValue::Map{};
                auto it = map.find("id");
                std::string id = (it != map.end() && it->second) ? it->second->toString() : "";
                cJSON *array = store.Load();
                cJSON *found = store.FindById(array, id);
                KRRenderValueMap result;
                if (found) {
                    result["conn"] = sftp_store::FromCJson(found);
                }
                cJSON_Delete(array);
                callback(KRRenderValue::Make(result));
                return;
            }

            if (method == METHOD_TOUCH_LAST_USED) {
                auto map = params ? params->toMap() : KRRenderValue::Map{};
                auto it = map.find("id");
                std::string id = (it != map.end() && it->second) ? it->second->toString() : "";
                cJSON *array = store.Load();
                cJSON *found = store.FindById(array, id);
                if (found) {
                    sftp_store::CJsonSetNumber(found, "lastUsedAt", sftp_store::NowMs());
                    store.SaveAndFree(array);
                } else {
                    cJSON_Delete(array);
                }
                callback(KRRenderValue::Make(KRRenderValueMap{}));
                return;
            }

            KRRenderValueMap err;
            err["error"] = KRRenderValue::Make(SftpErrorFormatter::Format(
                9999, "KRSftpConnectionModule unknown method: " + method));
            callback(KRRenderValue::Make(err));
        } catch (const std::exception &e) {
            KRRenderValueMap err;
            err["error"] = KRRenderValue::Make(SftpErrorFormatter::Format(e));
            callback(KRRenderValue::Make(err));
        }
    }).detach();
    return nullptr;
}

void KRSftpConnectionModule::OnDestroy() {
    // 连接列表持久化在文件中，无内存状态。
}

}  // namespace module
}  // namespace kuikly

File: core-render-android/src/main/java/com/tencent/kuikly/core/render/android/expand/module/KRSftpConnectionModule.kt
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.kuikly.core.render.android.expand.module

import android.content.Context
import com.tencent.kuikly.core.render.android.css.ktx.toJSONObjectSafely
import com.tencent.kuikly.core.render.android.css.ktx.toMap
import com.tencent.kuikly.core.render.android.export.KuiklyRenderBaseModule
import com.tencent.kuikly.core.render.android.export.KuiklyRenderCallback
import org.json.JSONArray
import org.json.JSONObject

/**
 * SFTP 连接列表 Module（Android 原生侧，§17.3.1 / §21.4.5）
 *
 * - 持久化到 SharedPreferences `sftp_connections`
 * - 删连接联动清收藏 + 历史（§21.4.5）
 * - 密码/密钥字段透传存储；Phase 1.2 接入 EncryptedSharedPreferences 加密
 */
class KRSftpConnectionModule : KuiklyRenderBaseModule() {

    override fun call(method: String, params: String?, callback: KuiklyRenderCallback?): Any? {
        return when (method) {
            METHOD_ADD -> add(params, callback)
            METHOD_UPDATE -> update(params, callback)
            METHOD_REMOVE -> remove(params, callback)
            METHOD_LIST -> list(params, callback)
            METHOD_GET -> get(params, callback)
            METHOD_TOUCH_LAST_USED -> touchLastUsed(params, callback)
            else -> super.call(method, params, callback)
        }
    }

    private fun add(params: String?, callback: KuiklyRenderCallback?) {
        executeOnSubThread {
            val json = params.toJSONObjectSafely()
            try {
                val storage = storage()
                val conn = json
                val id = conn.optString("id").ifEmpty { java.util.UUID.randomUUID().toString() }
                conn.put("id", id)
                conn.put("createdAt", System.currentTimeMillis())
                storage.save(conn)
                callback?.invoke(mapOf("id" to id))
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to SftpErrorFormatter.format(e)))
            }
        }
    }

    private fun update(params: String?, callback: KuiklyRenderCallback?) {
        executeOnSubThread {
            val json = params.toJSONObjectSafely()
            try {
                val id = json.optString("id")
                if (id.isEmpty()) throw IllegalArgumentException("missing id")
                val storage = storage()
                storage.save(json)  // 覆盖同 id
                callback?.invoke(mapOf("ok" to true))
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to SftpErrorFormatter.format(e)))
            }
        }
    }

    private fun remove(params: String?, callback: KuiklyRenderCallback?) {
        executeOnSubThread {
            val json = params.toJSONObjectSafely()
            try {
                val id = json.optString("id")
                val storage = storage()
                storage.remove(id)
                // 联动清收藏 + 历史（§21.4.5）——直接调原生 storage，避免 Module 异步回调
                SftpFavoritesStorage(context).removeByConnection(id)
                SftpPlaybackHistoryStorage(context).clearByConnection(id)
                callback?.invoke(mapOf("ok" to true))
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to SftpErrorFormatter.format(e)))
            }
        }
    }

    private fun list(params: String?, callback: KuiklyRenderCallback?) {
        executeOnSubThread {
            try {
                val storage = storage()
                val items = storage.loadAll()
                // 按 lastUsedAt DESC 排序
                val sorted = items.sortedByDescending { it.optLong("lastUsedAt", 0L) }
                // 裸 JSONArray 过不了桥，必须 JSONObject + toMap()（与 KRSftpModule.list 一致）
                val result = JSONObject()
                val arr = JSONArray()
                sorted.forEach { arr.put(it) }
                result.put("items", arr)
                callback?.invoke(result.toMap())
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to SftpErrorFormatter.format(e)))
            }
        }
    }

    private fun get(params: String?, callback: KuiklyRenderCallback?) {
        executeOnSubThread {
            val json = params.toJSONObjectSafely()
            try {
                val id = json.optString("id")
                val storage = storage()
                val conn = storage.findById(id)
                if (conn != null) {
                    val result = JSONObject()
                    result.put("conn", conn)
                    callback?.invoke(result.toMap())
                } else {
                    callback?.invoke(mapOf("error" to """{"code":3001,"msg":"connection not found"}"""))
                }
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to SftpErrorFormatter.format(e)))
            }
        }
    }

    private fun touchLastUsed(params: String?, callback: KuiklyRenderCallback?) {
        executeOnSubThread {
            val json = params.toJSONObjectSafely()
            try {
                val id = json.optString("id")
                val storage = storage()
                storage.touchLastUsed(id)
                callback?.invoke(mapOf("ok" to true))
            } catch (e: Exception) {
                callback?.invoke(mapOf("error" to SftpErrorFormatter.format(e)))
            }
        }
    }

    private fun storage(): SftpConnectionStorage = SftpConnectionStorage(context)

    private fun executeOnSubThread(block: () -> Unit) {
        com.tencent.kuikly.core.render.android.adapter.KuiklyRenderAdapterManager
            .krThreadAdapter?.executeOnSubThread(block) ?: Thread(block).start()
    }

    companion object {
        const val MODULE_NAME = "KRSftpConnectionModule"
        const val METHOD_ADD = "add"
        const val METHOD_UPDATE = "update"
        const val METHOD_REMOVE = "remove"
        const val METHOD_LIST = "list"
        const val METHOD_GET = "get"
        const val METHOD_TOUCH_LAST_USED = "touchLastUsed"
    }
}

/** SFTP 连接持久化（SharedPreferences） */
class SftpConnectionStorage(private val context: Context?) {
    private val prefs by lazy {
        context?.getSharedPreferences("sftp_connections", android.content.Context.MODE_PRIVATE)
    }

    fun save(conn: JSONObject) {
        val id = conn.optString("id")
        if (id.isEmpty()) return
        val arr = loadAll().filterNot { it.optString("id") == id }.toMutableList()
        arr.add(conn)
        prefs?.edit()?.putString(KEY, encodeAll(arr))?.apply()
    }

    fun remove(id: String) {
        val arr = loadAll().filterNot { it.optString("id") == id }
        prefs?.edit()?.putString(KEY, encodeAll(arr))?.apply()
    }

    fun findById(id: String): JSONObject? = loadAll().firstOrNull { it.optString("id") == id }

    fun loadAll(): List<JSONObject> {
        val raw = prefs?.getString(KEY, null) ?: return emptyList()
        return try {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { arr.optJSONObject(it) }.filterNotNull()
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun touchLastUsed(id: String) {
        val arr = loadAll().toMutableList()
        val idx = arr.indexOfFirst { it.optString("id") == id }
        if (idx < 0) return
        arr[idx].put("lastUsedAt", System.currentTimeMillis())
        prefs?.edit()?.putString(KEY, encodeAll(arr))?.apply()
    }

    private fun encodeAll(list: List<JSONObject>): String {
        val arr = JSONArray()
        list.forEach { arr.put(it) }
        return arr.toString()
    }

    companion object {
        private const val KEY = "items"
    }
}

/** KRSftpConnectionModule 全局单例 holder（§21.7.4） */
object KRSftpConnectionModuleHolder {
    val instance: KRSftpConnectionModule by lazy { KRSftpConnectionModule() }
}
